OpenTerm/Util/Cub.swift

```diff
@@ -0,0 +1,186 @@
+		fputs("Missing file \(path)\n", thread_stderr)
+		return 1
+	}
+
+	let runner = Runner(logDebug: true, logTime: false)
+
+	runner.registerExternalFunction(name: "print", argumentNames: ["input"], returns: true) { (arguments, callback) in
+
+		for (_, arg) in arguments {
+			
+			switch arg {
+			case .string(let str):
+				terminalView.performOnMain {
+					terminalView.appendText(NSAttributedString(string: str))
+				}
+			default:
+				break
+			}
+			
+		}
+
+		Thread.sleep(forTimeInterval: 0.02)
+		_ = callback(.number(1))
+		return
+	}
+
+	runner.registerExternalFunction(name: "exec", argumentNames: ["command"], returns: true) { (arguments, callback) in
+
+		var arguments = arguments
+
+		guard let command = arguments.removeValue(forKey: "command") else {
+			_ = callback(.number(1))
+			return
+		}
+
+		guard case let .string(commandStr) = command else {
+			_ = callback(.number(1))
+			return
+		}
+
+		executorCommand = CubCommandExecutor(commandStr: commandStr, terminalView: terminalView, callback: {
+			
+			DispatchQueue.main.async {
+				
+				_ = callback(.number(1))
+			}
+
+		})
+	
+	}
+
+	do {
+
+		runner.executionFinishedCallback = {
+
+			DispatchQueue.main.async {
+				
+				terminalView.stderrParser.delegate = terminalView
+				terminalView.stdoutParser.delegate = terminalView
+				
+				terminalView.executor.delegate = terminalView
+				
+				activeVC.terminalView.parserDidEndTransmission(activeVC.terminalView.stdoutParser)
+
+			}
+
+		}
+			
+		try runner.run(source)
+
+	} catch {
+		return 1
+	}
+
+	return 0
+}
+
+class CubCommandExecutor: CommandExecutorDelegate, ParserDelegate {
+	
+	let callback: (() -> Void)
+	
+	let terminalView: TerminalView
+	
+	init(commandStr: String, terminalView: TerminalView, callback: @escaping (() -> Void)) {
+		
+		self.callback = callback
+		self.terminalView = terminalView
+		
+		terminalView.stderrParser.delegate = self
+		terminalView.stdoutParser.delegate = self
+
+		terminalView.executor.delegate = self
+
+		terminalView.executor.dispatch(commandStr)
+	}
+	
+	func commandExecutor(_ commandExecutor: CommandExecutor, receivedStdout stdout: Data) {
+		terminalView.stdoutParser.parse(stdout)
+	}
+	
+	func commandExecutor(_ commandExecutor: CommandExecutor, receivedStderr stderr: Data) {
+		terminalView.stderrParser.parse(stderr)
+	}
+	
+	func commandExecutor(_ commandExecutor: CommandExecutor, didChangeWorkingDirectory to: URL) {
+		
+	}
+	
+	func commandExecutor(_ commandExecutor: CommandExecutor, stateDidChange newState: CommandExecutor.State) {
+		
+	}
+	
+	func parser(_ parser: Parser, didReceiveString string: NSAttributedString) {
+		
+		terminalView.performOnMain {
+			self.terminalView.appendText(string)
+		}
+		
+	}
+	
+	func parserDidEndTransmission(_ parser: Parser) {
+		
+		terminalView.performOnMain {
+
+			self.terminalView.stderrParser.delegate = nil
+			self.terminalView.stdoutParser.delegate = nil
+			
+			self.terminalView.executor.delegate = nil
```

Comment:
Not sure if it's safe to modify these delegates. This will have a conflict with PR #85, which moves the parsers down to a lower level.